src/v/iceberg/filesystem_catalog.h

```diff
@@ -0,0 +1,132 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+#pragma once
+
+#include "base/seastarx.h"
+#include "cloud_io/remote.h"
+#include "iceberg/catalog.h"
+#include "iceberg/partition.h"
+#include "iceberg/schema.h"
+#include "iceberg/table_io.h"
+#include "iceberg/table_metadata.h"
+#include "iceberg/transaction.h"
+#include "model/fundamental.h"
+#include "utils/named_type.h"
+
+#include <seastar/core/future.hh>
+#include <seastar/core/sstring.hh>
+
+namespace iceberg {
+
+// A simple catalog implementation that uses object storage to manage updates
+// to the table metadata. The file layout is similar to the Apache Iceberg
+// HadoopCatalog, which places metadata into per-table directories. This means
+// that metadata written with this catalog can be read by applications that
+// know how to read from the HadoopCatalog.
+//
+// Like the HadoopCatalog, this catalog advertises the latest metadata with a
+// "version hint", a file containing an integer that refers to the latest table
+// metadata. Note, this version is an artifact of the HadoopCatalog only, and
+// is not a part of the table metadata spec.
+//
+// The files laid out by this catalog look roughly like:
+//
+// /warehouse_location/my_namespace/my_table/metadata/version-hint.text
+//   - the version hint: contains the string "42"
+// /warehouse_location/my_namespace/my_table/metadata/v42.metadata.json
+//   - contains the latest table metadata as JSON
+// /warehouse_location/my_namespace/my_table/metadata/v41.metadata.json
+// /warehouse_location/my_namespace/my_table/metadata/v40.metadata.json
+// /warehouse_location/my_namespace/my_table/metadata/v...
+//   - contain older versions of the table metadata
+//   - Iceberg has the write.metadata.delete-after-commit.enabled property to
+//     delete these, though this is not yet implemented by filesystem_catalog
+//
+// Unlike the HadoopCatalog, which uses atomic renames to avoid overwriting
+// existing table metadata, this catalog is by comparison less safe, performing
+// best-effort non-atomic operations to discourage concurrent updates.
+//
+// WARNING: until the concurrency issues are addressed, this catalog should
+// only be used for testing and non-production workloads.
+//
+// WARNING: the file system based scheme to commit a metadata file is
+// documented as deprecated and will eventually be phased out by the Apache
+// Iceberg ecosystem.
```

Comment:
how should i reconcile these two warnings? they seem to be in conflict, if read perhaps in the future after full deprecation by apache iceberg?

Reply:
Hm maybe I should clarify these. I mean to say that 1) it's unsafe without atomic operations, and that 2) the ecosystem strongly recommends using a real catalog service, rather than the file system. They're both statements to steer people towards using a real catalog like Polaris.

Reply:
Also FWIW we can address 1) and ignore the recommendation from the ecosystem -- the deprecation came before AWS announced conditional writes for S3, and IMO it's not unlikely people will continue to want to use this style of operation (especially since not every query engine supports REST catalogs)

Reply:
makes sense thx